### DISCUSSION TITLE
Offer fine grained control for code mutation

### DISCUSSION THREAD

**Discussion by dupdob** [stryker-mutator/stryker-net#1503] [Ideas]
Stryker.net could provide fine grained control for source code mutation via code comments, eg:
```csharp
...
  if (IsActive(someThing))
  {
    // NO MUTATION
    SomeMethod(a , b, true);
  }
...
```
would prevent mutations of the `SomeMethod(a, b, true)` line.

The comment could have to respect some predefined format to provide even finer control, e.g:
 - provide a rationale: `// NO MUTATION (Unkillable mutant)`
 - disable specific mutators `// NO MUTATION [String, BinaryExpression]`
 - provide scoped control: 
```csharp
 ...
// NO MUTATION: start
...
// NO MUTATION: end
...
```
- this could be controlled by a command line options to force mutations (or maybe only to mutate those lines), to allow teams to perform periodic reviews. This override could be automated in diff mode, when there is change in a `no mutation` line

**Comment by psfinaki** [comment]
Adding comments solely for testing purpose sounds like a hack for me. A developer unfamiliar with the mutation testing concept will be confused looking at that. 

Another problem will be with static code analyzers that tend to issue warnings and errors like "remove this unneeded comment" so they would also have to be adjusted - imagine this in the code for every case:
```
#pragma warning disable CS8625
// NO MUTATION: start
#pragma warning restore

actual code

#pragma warning disable CS8625
// NO MUTATION: end
#pragma warning restore

**Reply by dupdob** [reply]
I am not sure why you talk about 'for testing purpose', definitely not my intent in any way.
Here is my problem:
- I want to achieve 100% mutant coverage, I am currently at 99.38% (as in 13 survivors out of 2053) (why is out of the scope of this discussion)
- to reach that score, I added some tests, and refactored my code to make it leaner
- but sometimes, the resulting was code felt ugly or lost much in readability
- and those 13 survivors are un-killable by construction because they do not significantly alter the code behavior

I would love to be able to disable those survivors as well as those other mutants that are not worth to be killed.

**Reply by dupdob** [reply]
Regarding the unneeded comment, I would expect a real use case to look like this:
```csharp
...
  if (IsActive(options2))
  {
    // Mutating this line will result in un killable mutants. Remove the comment and check if the code is modified
    // NO MUTATION
    activeOptions |= Parse.Option2;
  }
...
```

**Reply by dupdob** [reply]
Regarding confusion: this is probably a matter of message format, I agree that the proposal is pretty short and basic, it could be made longer for it to be more expressive, such as:
```
  // Stryker Specific: NO MUTATION
```

**Reply by psfinaki** [reply]
Well by "for testing purposes" I meant "for mutation testing purposes", mutation testing is testing anyway. 

I get the idea, however I consider this mutation score similar to test coverage percentage - it's good to aim for the absolute value but getting there is often not realistic or not worth the cost. So if the redesign of code is too expensive, just put up with the numbers :)

I mean such things technically don't hurt when they are "opt-in" but I imagine them being abused as well - by those lazy guys who just want to shut up the technology instead of dealing with the issue.

Anyway - this is just my opinion on that as a developer, the maintainers might think otherwise.

**Reply by dupdob** [reply]
Thanks for the clarification. An alternative suitable solution for me would be to adjust offending mutator so that they dont end up creating un-killable mutants.
Another good solution would be for me to mark those mutants as : 'reviewed, please ignore'. This one could be better because the flag could be automatically erased when the code changes.

The point of reaching 100% is that every survivor needs to be reviewed, analyzed and fixed. As of now, if I have a new survivors (number 14), I need to review them to find the new one and revise it.

As the principal maintainer for [NFluent](https://github.com/NFluent/NFluent), I have to ensure the assertion framework is defect free, as it is a testing tool. I know this is an edge case, but in my view, Mutation Testing is mostly valuable when you reach high test coverage

**Reply by rouke-broersma** [reply]
We mostly share the same view as @psfinaki , however time and time again users ask for this feature. Over time we have implemented some limited ways to exclude code from mutation. Perhaps it is time to get over it and add fully implement the feature if only so those users that want it can stop asking for it. I believe our friends over at stryker-mutator/stryker-js have implemented this feature as well. I would prefer if our comment style is similar to theirs (while still feeling native for a dotnet developer, so exact match is not necessary).

**Reply by sbergen** [reply]
I'm on the same train as @dupdob, being the sole maintainer of [Responsible](https://github.com/sbergen/Responsible) (it's a side project for me, but we use it at work). I've been using `[ExcludeFromCodeCoverage]` for assertion methods that should never be triggered through the public API (which is all I test, it's been a very functional approach for me), and currently have both 100% coverage and 100% mutant score. This way I can be sure that nothing never falls out of coverage, as with a threshold you might add some here, and remove some there, and still stay below it.

However, I was just testing the v1.0 branch, and I'll need to do some some work on improving a few places, but there are some places that don't make sense to kill. So I'd definitely like to see something like this also.

I wasn't able to find any documentation for the JS comments, so I don't know how they compare. But in any case, ReSharper uses a similar thing, where comments can look like
```
// ReSharper disable All
// ReSharper disable PossibleNullReferenceException
// ReSharper restore PossibleNullReferenceException
// ReSharper disable once PossibleNullReferenceException
```

The nice thing is that they allow you to add other stuff on the same line, so you can do
```
// ReSharper disable once foo PossibleNullReferenceException bar
```
and it still works. A more valid use case is something like
```
// ReSharper disable once PossibleNullReferenceException, used inside try-catch
```

So I would suggest a similar format, where
* if a comment starts with `Stryker`, it's considered a mutation control comment
* the next word must be `disable` or `restore`
* If the previous word was disable, the next word may be `once` to apply only to the next line
* After this, all words are parsed
  * `All` means all mutators
  * If a word matches a mutator, it is affected
  * If a word does not match a mutator, it is simply ignored

**Reply by rouke-broersma** [reply]
I thought our javascript friends had implemented it by now, seems that they haven not. Their proposal is here: https://github.com/stryker-mutator/stryker-js/issues/1472#proposed-implementation 

Your suggestions are very similar, I like it.

Keep in mind that this feature is completely non-trivial and that I personally have little interest in implementing it at this time.

**Reply by dupdob** [reply]
I also like the syntax. But it is somewhat different from the JS proposal (`next` which does not really makes to me).
Assuming this difference an issue, I think I could implement this, here are the envisioned specs:

- [x] works on whole statements at minimum: this is due to how Roslyn attaches comments and code. Will do finer grained control (e.g. expression) if possible, but not a priority
- [x] works on unitary/automatic scope as a priority, manually defined scope (i.e using `disable` and `restore`) is a secondary objective. This derives from the assumption this feature should be used sparingly and not in broad strokes.
- [x] mutations will still be generated but will be flagged as ignored. First implementations may simply skip generating mutations. Remark: skipping mutating altogether may provide a workaround for when a mutator triggers an exception blocking all usage of Stryker  ⇒ we need to choose which behavior we want.
- [x] works at statement, block or declaration level. The rule then apply to the attached scope. eg
```csharp
{
  // Stryker disable once all
  int x = 2+2;
  int y = x+3;
}
```
 ⇒ disable mutation for `int x = 2+2;` 
 ```csharp
// Stryker disable once all
{
  int x = 2+2;
  int y = x+3;
}
```
 ⇒ disable mutation for both lines
 - [x] manual defined scope (i.e using `disable` and `restore`) does not work accross scope (i.e. you cannot disable mutations in one method and reenable it somewhere in the next method). This is due to the tree parsing logic we use which makes this needlessly difficult to implement; plus it should be used sparingly
 - [ ] full comment syntax should be
    `// Stryker [disable[| once]|restore] [all|mutator1[|,mutator2...]][| : reason for disabling]`
    the `'reason for disabling` will be used to comment skipped mutations (assuming mutations are stil generated, see above)  ⇒ we need to decide if we want to make this part mandatory or not. I am in favor of making it mandatory
- [ ] Stryker could still generates mutations in `diff-mode` when the lines in the scope were mutated, giving an opportunity to review those lines. Note this probably means significant effort, as it means accessing diff information during mutation phase. Probably a feature for a future version, if possible at all.
- [ ] Stryker could issue a warning, or block altogether if ratio of skipped mutants is above a user defined threshold. This may be useful for teams to control overall discipline with the feature.

**Reply by rouke-broersma** [reply]
> manual defined scope (i.e using `disable` and `restore`) does not work accross scope (i.e. you cannot disable mutations in one method and reenable it somewhere in the next method). This is due to the tree parsing logic we use which makes this needlessly difficult to implement; plus it should be used sparingly

Even if we could, I do not think we would want to support this. Does not make sense to me. Disabling always only means something for the scope it is defined in at the most and optionally only for the nearest scope (with `once`).

> full comment syntax should be
>   `// Stryker [disable[| once]|restore] [all|mutator1[|,mutator2...]][| : reason for disabling]`
>   the `'reason for disabling` will be used to comment skipped mutations (assuming mutations are stil generated, see above)  ⇒ we need to decide if we want to make this part mandatory or not. I am in favor of making it mandatory

I don't have an opinion on making it mandatory one way or another. If we don't make it mandatory we can use a default like we currently do for IgnoreCodeCoverageAttribute etc.

All in all I agree with the proposed solution. Perhaps you can move this to an issue (if we don't yet have one) and we can close this discussion.

**Reply by dupdob** [reply]
I will copy past my latest comment to a new issue. I suspect a similar one already exists, but a cursory search revealed nothing, so at worst will link any old one the new one
